File: src/core/domain/intent.ts
/**
 * Intent — 创作意图（v3 新增 · Stage 12.1）
 *
 * 设计原则：
 * - 显式抽象「用户想做哪种解说」，驱动 L0-L2 各阶段的策略选择
 *   （如 short-drama 偏钩子+反转，movie-review 偏深度分析）。
 * - 替代 v2 隐式的「5 种 ScriptStylePreset」风格枚举：v3 把风格
 *   收敛到 IntentConfig.toneIntensity 一个 0.0-1.0 维度，不再是离散枚举。
 * - 数据流：用户在 workspace/pick-intent.tsx 选 intent → 写入 Project.intent
 *   → 流水线各 step 读取 project.intent 调整 prompt / 段落数 / 风格。
 *
 * v3 仅实现 zh-CN（中文）；en-US / ja-JP 留 enum 占位，前端 v3.1 补 UI。
 */

import type { ScriptStylePreset } from '@/types/script';

/** 内容意图（v3 枚举） */
export type ContentIntent =
  | 'movie-review' // 电影解说：深度分析 + 剧情梳理
  | 'short-drama' // 短剧解说：钩子 + 反转 + 吐槽
  | 'comic-drama' // 漫剧解说：v3 数据层预留，UI v3.1 补
  | 'episode-recap' // 剧集回顾：按集
  | 'voice-over' // 纯配音：v3 数据层预留，UI v3.1 补
  | 'highlight' // 高光集锦：复用 v2 剪辑模式能力
  | 'auto'; // AI 自动判断（基于视频元数据）

/** 支持的语言（v3 仅 zh-CN，其余占位） */
export type IntentLanguage = 'zh-CN' | 'en-US' | 'ja-JP';

/** 目标受众 */
export type IntentAudience = 'general' | 'professional' | 'young';

/** 意图配置（写入 Project.intent） */
export interface IntentConfig {
  intent: ContentIntent;
  /** 目标解说时长（秒）。短剧 60-180，电影 180-600，episode-recap 180-300。 */
  targetDurationSecs: number;
  language: IntentLanguage;
  audience: IntentAudience;
  /** 语气强度：0 = 克制，1 = 强烈（替代 v2 离散风格枚举） */
  toneIntensity: number;
  /** 可选：v2 风格预设映射（保留向后兼容字段，v3.1 移除） */
  legacyStyle?: ScriptStylePreset;
}

// ─── 常量 ──────────────────────────────────────────────────

/** 默认配置（普通解说 · 中文 · 通用受众 · 中性语气） */
export const DEFAULT_INTENT_CONFIG: IntentConfig = {
  intent: 'short-drama',
  targetDurationSecs: 180,
  language: 'zh-CN',
  audience: 'general',
  toneIntensity: 0.5,
};

/** 各意图的默认目标时长（秒） */
export const DEFAULT_DURATION_BY_INTENT: Record<ContentIntent, number> = {
  'movie-review': 300,
  'short-drama': 120,
  'comic-drama': 120,
  'episode-recap': 240,
  'voice-over': 60,
  highlight: 60,
  auto: 180,
};

// ─── 纯函数 ──────────────────────────────────────────────

/** 校验意图字符串是否合法 */
export function isValidIntent(s: string): s is ContentIntent {
  return [
    'movie-review',
    'short-drama',
    'comic-drama',
    'episode-recap',
    'voice-over',
    'highlight',
    'auto',
  ].includes(s);
}

/** 根据 intent 推导默认 IntentConfig（其余字段为默认） */
export function intentDefaultConfig(intent: ContentIntent): IntentConfig {
  return {
    ...DEFAULT_INTENT_CONFIG,
    intent,
    targetDurationSecs: DEFAULT_DURATION_BY_INTENT[intent],
  };
}

/** toneIntensity 0-1 转 5 档风格枚举（兼容 v2 ScriptStylePreset） */
export function intensityToStyle(intensity: number): ScriptStylePreset {
  if (intensity < 0.2) return 'serious';
  if (intensity < 0.4) return 'conversational';
  if (intensity < 0.6) return 'warm';
  if (intensity < 0.8) return 'humorous';
  return 'suspense';
}

File: src-tauri/src/domain/intent.rs
//! Intent — 创作意图（与前端 `src/core/domain/intent.ts` 对齐）
//!
//! v3 新增抽象：替代 v2 离散 ScriptStylePreset 风格枚举，
//! 显式声明「用户想做哪种解说」，驱动 L0-L2 各阶段策略。
//! toneIntensity 0.0-1.0 连续值表达语气强度。

use serde::{Deserialize, Serialize};

// ─── 枚举 ──────────────────────────────────────────────────────

/// 内容意图（v3 枚举，7 值）
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ContentIntent {
    /// 电影解说：深度分析 + 剧情梳理
    MovieReview,
    /// 短剧解说：钩子 + 反转 + 吐槽
    ShortDrama,
    /// 漫剧解说：v3 数据层预留
    ComicDrama,
    /// 剧集回顾：按集
    EpisodeRecap,
    /// 纯配音：v3 数据层预留
    VoiceOver,
    /// 高光集锦：复用 v2 剪辑模式能力
    Highlight,
    /// AI 自动判断
    Auto,
}

/// 支持的语言（v3 仅 zh-CN，其余占位）
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum IntentLanguage {
    #[serde(rename = "zh-CN")]
    ZhCn,
    #[serde(rename = "en-US")]
    EnUs,
    #[serde(rename = "ja-JP")]
    JaJp,
}

/// 目标受众
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum IntentAudience {
    General,
    Professional,
    Young,
}

// ─── 意图配置 ──────────────────────────────────────────────────

/// 意图配置（写入 Production.intent）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IntentConfig {
    pub intent: ContentIntent,
    /// 目标解说时长（秒）
    pub target_duration_secs: f64,
    pub language: IntentLanguage,
    pub audience: IntentAudience,
    /// 语气强度 0.0-1.0
    pub tone_intensity: f32,
}

// ─── 常量 ──────────────────────────────────────────────────────

/// 默认配置（普通解说 · 中文 · 通用受众 · 中性语气）
pub const DEFAULT_INTENT_CONFIG: IntentConfig = IntentConfig {
    intent: ContentIntent::ShortDrama,
    target_duration_secs: 180.0,
    language: IntentLanguage::ZhCn,
    audience: IntentAudience::General,
    tone_intensity: 0.5,
};

/// 各意图的默认目标时长（秒）
pub fn default_duration_by_intent(intent: ContentIntent) -> f64 {
    match intent {
        ContentIntent::MovieReview => 300.0,
        ContentIntent::ShortDrama => 120.0,
        ContentIntent::ComicDrama => 120.0,
        ContentIntent::EpisodeRecap => 240.0,
        ContentIntent::VoiceOver => 60.0,
        ContentIntent::Highlight => 60.0,
        ContentIntent::Auto => 180.0,
    }
}

// ─── 纯函数 ──────────────────────────────────────────────────

/// 校验意图字符串是否合法（前端传来的字符串反序列化时使用）
pub fn is_valid_intent(s: &str) -> bool {
    matches!(
        s,
        "movie-review"
            | "short-drama"
            | "comic-drama"
            | "episode-recap"
            | "voice-over"
            | "highlight"
            | "auto"
    )
}

/// 根据 intent 推导默认 IntentConfig
pub fn intent_default_config(intent: ContentIntent) -> IntentConfig {
    IntentConfig {
        target_duration_secs: default_duration_by_intent(intent),
        ..DEFAULT_INTENT_CONFIG
    }
}

/// toneIntensity 0-1 转 5 档风格枚举（与前端 ScriptStylePreset 对齐）
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ScriptStyle {
    Serious,
    Conversational,
    Warm,
    Humorous,
    Suspense,
}

pub fn intensity_to_style(intensity: f32) -> ScriptStyle {
    if intensity < 0.2 {
        ScriptStyle::Serious
    } else if intensity < 0.4 {
        ScriptStyle::Conversational
    } else if intensity < 0.6 {
        ScriptStyle::Warm
    } else if intensity < 0.8 {
        ScriptStyle::Humorous
    } else {
        ScriptStyle::Suspense
    }
}
